compiler/ast.rs

```diff
@@ -0,0 +1,633 @@
+    tsx,
+    ..TsConfig::default()
+  }
+}
+
+fn get_syntax(media_type: MediaType) -> Syntax {
+  match media_type {
+    MediaType::JavaScript => Syntax::Es(get_es_config(false)),
+    MediaType::JSX => Syntax::Es(get_es_config(true)),
+    MediaType::TypeScript => Syntax::Typescript(get_ts_config(false)),
+    MediaType::TSX => Syntax::Typescript(get_ts_config(true)),
+    _ => Syntax::Es(get_es_config(false)),
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct TranspileOptions {
+  emit_metadata: bool,
+  jsx_factory: String,
+  jsx_factory_fragment: String,
+  transform_jsx: bool,
+}
+
+impl Default for TranspileOptions {
+  fn default() -> Self {
+    TranspileOptions {
+      emit_metadata: false,
+      jsx_factory: "React.createElement".into(),
+      jsx_factory_fragment: "React.Fragment".into(),
+      transform_jsx: true,
+    }
+  }
+}
+
+#[derive(Clone)]
+pub struct ParsedModule {
+  comments: SingleThreadedComments,
+  module: Module,
+  source_map: Rc<SourceMap>,
+}
+
+impl fmt::Debug for ParsedModule {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "ParsedModule {{ }}")
+  }
+}
+
+impl ParsedModule {
+  pub fn get_leading_comments(&self, span: &Span) -> Vec<Comment> {
+    self
+      .comments
+      .with_leading(span.lo, |comments| comments.to_vec())
+  }
+
+  /// Identifies all syntactical dependencies of a module (e.g. `import` and
+  /// `export` statements).
+  pub fn get_dependencies(&self) -> Vec<DependencyDescriptor> {
+    let mut collector = DependencyCollector(Vec::new());
+    collector.visit_module(&self.module, &self.module);
+
+    collector.0
+  }
+
+  pub fn get_location(&self, span: &Span) -> Location {
+    self.source_map.lookup_char_pos(span.lo).into()
+  }
+
+  pub fn get_module_leading_comments(&self) -> Vec<Comment> {
+    self.get_leading_comments(&self.module.span)
+  }
+
+  pub fn transpile(&self, options: TranspileOptions) -> Result<String> {
+    let program = Program::Module(self.module.clone());
+    let transform_jsx = options.transform_jsx;
+    let jsx_pass = react::react(
+      self.source_map.clone(),
+      react::Options {
+        pragma: options.jsx_factory,
+        pragma_frag: options.jsx_factory_fragment,
+        // this will use `Object.assign()` instead of the `_extends` helper
+        // when spreading props.
+        use_builtins: true,
+        ..Default::default()
+      },
+    );
+    let mut compiler_pass = chain!(
+      decorators::decorators(decorators::Config {
+        legacy: true,
+        emit_metadata: options.emit_metadata,
+      }),
+      helpers::InjectHelpers,
+      typescript::strip(),
+      fixer(Some(&self.comments)),
+    );
+    let program = swc_common::GLOBALS.set(&Globals::new(), || {
+      helpers::HELPERS.set(&helpers::Helpers::new(false), || {
+        if transform_jsx {
+          program.fold_with(&mut chain!(jsx_pass, compiler_pass))
+        } else {
+          program.fold_with(&mut compiler_pass)
+        }
```

Comment:
Nitpick, move conditional outside of nested closures?

Reply:
I would too, but the problem is that `chain!` produces these really complex opaque types (basically chained `AndThen` types) that I can seem to find a type I can assign a variable that allows me to then pass it outside the closure.  Maybe it is the limits of my Rust capabilities, but I can't seem to find something that works.

Reply:
You can workaround this by using `either::Either`.

I use `Either::Left(noop())` and `Either::Right(my_pass)` when I don't want to box it.

Reply:
awesome, thank you, that works well...